presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java

```diff
@@ -65,6 +65,7 @@ public JoinStatsRule(FilterStatsCalculator filterStatsCalculator, StatsNormalize
     {
         super(normalizer);
         this.filterStatsCalculator = requireNonNull(filterStatsCalculator, "filterStatsCalculator is null");
+        this.normalizer = normalizer;
```

Comment:
It cannot assess the impact of this change. I know there are tests and the tests have their expectations appropriately updated, but these are synthetic tests and are as good as they expected values.

Would it be possible to bring more confidence to this change by adding a test in one of the `TestTpc*Stats` classes?

Reply:
Could you please elaborate more on this?

Reply:
We have many tests for JOIN in `TestTpc*Stats`. What additional tests would you like to see?

Reply:
Hm.. but none of these needed (nor enjoyed) an update. So there is nothing ensuring we don't regress in the future.

Reply:
Tests in `TestTpc*Stats` are not checking for exact estimate, and allow deviations.

Reply:
I know that. If the change improves the estimations, "allowed deviations" could be reduced.
If this is not possible (or does not make sense; or anyhow would not ensure we do not regress), then we need an explicit test.

Reply:
I'm not sure if normalization improves accuracy. That's not the point of this commit. JoinStatsRule is logic intensive. A lot of arithmetic is being done. The FilterStatsCalculator is used in multiple places. As everywhere else across the code (including `FilterStatsCalculator`) it is assumed that stats are normalized. (e.g. rowCount >= ndv, ndv consistent with range, nulls factor consistent with ndv, etc.) it seems to be just safer to do not break this assumption in the `JoinStatsRule` as well.

Reply:
Fair enough. Keep in mind that join stats rule is executed in tight loop within join enumeration so it has to be performant therefore we should use normalization only where it has chance to fix something. 

@findepi do you recall when normalization made plans worse?

Reply:
> @findepi do you recall if/when normalization made plans worse?

unfortunately i don't remember. It was introduced quite some time ago, to avoid totally inconsistent stats (eg. NDV > row count after a filter for symbols not involved in the filter condition), as they made further rules produce stupid results (like `x = c` producing < 1 rows, even though we assume this should produce 1+).

Reply:
@sopel39 This is a very valid concert. I'm running the benchmarks to estimate the impact.

Reply:
Benchmarks didn't show any significant (visible) regressions after applying normalization.

Before:

```
# Run complete. Total time: 00:02:18

Benchmark                                           (joinReorderingStrategy)  Mode  Cnt    Score   Error  Units
BenchmarkReorderChainedJoins.benchmarkReorderJoins     ELIMINATE_CROSS_JOINS  avgt   30   19.224 ± 0.626  ms/op
BenchmarkReorderChainedJoins.benchmarkReorderJoins                 AUTOMATIC  avgt   30  215.899 ± 4.233  ms/op

Benchmark                                                  (joinReorderingStrategy)  (numberOfTables)  Mode  Cnt      Score     Error  Units
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins     ELIMINATE_CROSS_JOINS                 2  avgt   30      6.109 ±   0.315  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins     ELIMINATE_CROSS_JOINS                 4  avgt   30      9.982 ±   0.478  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins     ELIMINATE_CROSS_JOINS                 6  avgt   30     13.560 ±   0.381  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins     ELIMINATE_CROSS_JOINS                 8  avgt   30     17.579 ±   0.425  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins     ELIMINATE_CROSS_JOINS                10  avgt   30     22.274 ±   0.537  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins                 AUTOMATIC                 2  avgt   30      7.378 ±   0.442  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins                 AUTOMATIC                 4  avgt   30     19.342 ±   0.675  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins                 AUTOMATIC                 6  avgt   30    116.208 ±   4.158  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins                 AUTOMATIC                 8  avgt   30   1056.741 ±  28.317  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins                 AUTOMATIC                10  avgt   30  10150.088 ± 149.229  ms/op

```

After:

```
Benchmark                                           (joinReorderingStrategy)  Mode  Cnt    Score   Error  Units
BenchmarkReorderChainedJoins.benchmarkReorderJoins     ELIMINATE_CROSS_JOINS  avgt   30   19.305 ± 0.552  ms/op
BenchmarkReorderChainedJoins.benchmarkReorderJoins                 AUTOMATIC  avgt   30  217.149 ± 5.105  ms/op

# Run complete. Total time: 00:21:16

Benchmark                                                  (joinReorderingStrategy)  (numberOfTables)  Mode  Cnt      Score      Error  Units
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins     ELIMINATE_CROSS_JOINS                 2  avgt   30      6.221 ±    0.320  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins     ELIMINATE_CROSS_JOINS                 4  avgt   30      9.663 ±    0.375  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins     ELIMINATE_CROSS_JOINS                 6  avgt   30     13.216 ±    0.423  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins     ELIMINATE_CROSS_JOINS                 8  avgt   30     17.506 ±    0.461  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins     ELIMINATE_CROSS_JOINS                10  avgt   30     22.591 ±    0.823  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins                 AUTOMATIC                 2  avgt   30      7.491 ±    0.554  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins                 AUTOMATIC                 4  avgt   30     20.584 ±    0.829  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins                 AUTOMATIC                 6  avgt   30    116.934 ±    3.790  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins                 AUTOMATIC                 8  avgt   30   1105.047 ±   28.931  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins                 AUTOMATIC                10  avgt   30  10309.552 ±  153.314  ms/op
```

Reply:
It actually makes sense. The normalization is very cheap if it does nothing.